File: Day Three 22-10-24/solution-1.js
// Letter Combinations of a Phone Number:
// Given a string containing digits from 2-9 inclusive, return all possible letter combinations 
// that the number could represent. Return the answer in any order.
//
// A mapping of digits to letters (just like on the telephone buttons) is given below. 
// Note that 1 does not map to any letters.
//
// Constraints:
// 0 <= digits.length <= 4
// digits[i] is a digit in the range ['2', '9'].


// The function `letterCombinations` generates all possible letter combinations 
// that the number could represent on a phone keypad, given a string of digits.
// If the input `digits` is empty, it returns an empty array.

const letterCombinations = (digits) => {
    if (!digits.length) return []; // Return an empty array if no digits are provided.

    // Mapping of digits to corresponding letters on a phone keypad.
    const phoneMap = {
        '2': 'abc',
        '3': 'def',
        '4': 'ghi',
        '5': 'jkl',
        '6': 'mno',
        '7': 'pqrs',
        '8': 'tuv',
        '9': 'wxyz'
    };

    const result = []; // Array to store the resulting combinations.

    // Helper function to perform backtracking.
    function backtrack(index, path) {
        // If the current path length equals the digits length, a valid combination is formed.
        if (index === digits.length) {
            result.push(path); // Add the current combination to the result array.
            return; // Exit the function.
        }

        // Get the letters corresponding to the current digit.
        const letters = phoneMap[digits[index]];
        // Iterate over each letter and continue building the combination.
        for (let letter of letters) {
            backtrack(index + 1, path + letter); // Recursive call to build the next level of the combination.
        }
    }

    backtrack(0, ''); // Start the backtracking process with an empty path.
    return result; // Return the array of all possible combinations.
}

File: Day Three 22-10-24/solution-1.py
def letter_combinations(digits: str):
    # Return an empty list if the input string is empty
    if not digits:
        return []
    
    # Mapping of digits to corresponding letters on a phone keypad
    phone_map = {
        '2': 'abc',  # Digit '2' maps to 'a', 'b', 'c'
        '3': 'def',  # Digit '3' maps to 'd', 'e', 'f'
        '4': 'ghi',  # Digit '4' maps to 'g', 'h', 'i'
        '5': 'jkl',  # Digit '5' maps to 'j', 'k', 'l'
        '6': 'mno',  # Digit '6' maps to 'm', 'n', 'o'
        '7': 'pqrs', # Digit '7' maps to 'p', 'q', 'r', 's'
        '8': 'tuv',  # Digit '8' maps to 't', 'u', 'v'
        '9': 'wxyz'  # Digit '9' maps to 'w', 'x', 'y', 'z'
    }
    
    # List to store the resulting combinations
    result = []
    
    # Helper function to perform backtracking
    def backTrack(index, path):
        # If the current path length equals the digits length, a valid combination is formed
        if index == len(digits):
            result.append(path)  # Add the current combination to the result list
            return
        
        # Get the letters corresponding to the current digit
        letters = phone_map[digits[index]]
        
        # Iterate over each letter and continue building the combination
        for letter in letters:
            backTrack(index + 1, path + letter)  # Recursive call to build the next level of the combination
            
    # Start the backtracking process with an empty path
    backTrack(0, '')
    
    # Return the list of all possible combinations
    return result